File: 0684-redundant-connection/0684-redundant-connection.cpp
class Solution {
public:
    
    vector<int> findRedundantConnection(vector<vector<int>>& edges) {
        if(edges.size()==0) return{};
        vector<int> parent;
        parent.resize(edges.size()+1);
        for(int i=1;i<edges.size()+1;i++){
            parent[i]=i;
        }
        for(vector<int> edge : edges){
            int f1=find(edge[0],parent);
            int f2=find(edge[1],parent);
            if(f1!=f2)
                parent[f1]=f2;
            else
                return edge;
        }
        return {};
    }
    int find(int x,vector<int> parent){
        return parent[x]==x ? x : find(parent[x],parent);
    }
    
};

File: 0684-redundant-connection/0684-redundant-connection.java
class Solution {
    public int[] findRedundantConnection(int[][] edges) {
        int n = edges.length;
        int[] parent = new int[n+1];
        Arrays.fill(parent, -1);
        for(int[] edge: edges){
            int x = getParent(parent, edge[0]);
            int y = getParent(parent, edge[1]);
            if(x == y){
                return edge;
            }
            parent[y]= x;
        }
        return new int[]{0,0};
    }
    public int getParent(int[] parent, int i){
        if(parent[i]==-1) return i;
        return parent[i]=getParent(parent, parent[i]);
    }
}